cpp/src/arrow/device.cc

```diff
@@ -193,6 +193,9 @@ Result<std::shared_ptr<Buffer>> CPUMemoryManager::ViewBufferFrom(
   if (!from->is_cpu()) {
     return nullptr;
   }
+  if (buf->device_type() != DeviceAllocationType::kCPU) {
+    return std::make_shared<Buffer>(buf->address(), buf->size(), shared_from_this());
```

Comment:
The problem is that the returned buffer won't keep the underlying `buf` alive. You should set `buf` as parent.

Reply:
I meant to do that and forgot, good catch! I've updated this and the other spot accordingly